### PR TITLE
FLUID-6620: convert Textfield Stepper to number input

### PR DESCRIPTION
**Pull request by greatislander**
This PR resolves [FLUID-6620](https://issues.fluidproject.org/browse/FLUID-6620) by updating the Textfield Stepper component throughout Infusion to a standard `<input type="number>` ([MDN docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number)). This resolves the perplexing behaviour in VoiceOver on Safari where announced values are represented as a decimal based on `aria-valuenow` / `aria-valuemax` for inputs with the `spinbutton` role instead of the displayed value (see: https://bugs.webkit.org/show_bug.cgi?id=221102).

## Additional Information

The `number` input type has native `stepUp()` and `stepDown()` methods for incrementing and decrementing its value. It might be good to leverage these somehow instead of using the custom logic defined in `fluid.textfieldStepper.step`, but that's a nice to have.

### CODE REVIEW

File: src/components/textfieldControl/css/TextfieldStepper.css [hunk 1]
```diff
@@ -23,13 +23,20 @@
 /* Textfield Styling */
 
 .fl-textfieldStepper .fl-textfieldStepper-textField {
+    -moz-appearance: textfield;
```
Comment (amb26) [inline src/components/textfieldControl/css/TextfieldStepper.css:26]: Which current browsers does this tweak display in? Is there a route to eliminating it someday?

File: src/components/textfieldControl/js/TextfieldStepper.js [hunk 2]
```diff
@@ -74,18 +79,13 @@ https://github.com/fluid-project/infusion/raw/main/Infusion-LICENSE.txt
...
-                            method: "attr",
-                            args: ["role", "spinbutton"]
                         }
                     },
                     modelListeners: {
                         "value": {
                             "this": "{that}.container",
                             method: "attr",
-                            args: ["aria-valuenow", "{change}.value"]
+                            args: ["value", "{change}.value"]
```
Comment (amb26) [inline src/components/textfieldControl/js/TextfieldStepper.js:88]: It seems from the MDN docs that this attribute controls a default rather than a current value - https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number#value - what's the accessibility impact of removing this listener?

Comment (amb26) [reply src/components/textfieldControl/js/TextfieldStepper.js:88]: Sorry, I should more properly say, "replacing this listener with one that changes the actual control value", via some means like $.val() or the direct "value" property, etc.

Comment (jobara) [reply src/components/textfieldControl/js/TextfieldStepper.js:88]: This should already be handled by the textfield component. https://github.com/fluid-project/infusion/blob/75c3c7e12a96f0ed168e8e2c9691ce5b061309c3/src/components/textfieldControl/js/Textfield.js#L44-L50

File: src/framework/preferences/css/sass/PrefsEditor.scss [hunk 3]
```diff
@@ -193,12 +193,19 @@ $icon-font: 'PrefsFramework-Icons';
...
         text-align: center;
         width: 4em;
     }
 
+    .fl-textfieldStepper .fl-textfieldStepper-textField::-webkit-inner-spin-button,
+    .fl-textfieldStepper .fl-textfieldStepper-textField::-webkit-outer-spin-button {
+        appearance: none;
+        margin: 0;
+    }
+
```
Comment (jobara) [inline src/framework/preferences/css/sass/PrefsEditor.scss:208]: Shouldn't these new additions already be coming from TextfieldStepper.css?

File: tests/component-tests/textfieldControl/js/TextfieldStepperTests.js [hunk 4]
```diff
@@ -89,7 +89,7 @@ https://github.com/fluid-project/infusion/raw/main/Infusion-LICENSE.txt
 
         fluid.tests.textfieldStepper.verifyState = function (that, expectedValue, incBtnState, decBtnState) {
             jqUnit.assertEquals("The model value should be " + expectedValue, expectedValue, that.model.value);
-            jqUnit.assertEquals("The aria-valuenow should be " + expectedValue, expectedValue.toString(), that.textfield.container.attr("aria-valuenow"));
+            jqUnit.assertEquals("The value should be " + expectedValue, expectedValue.toString(), that.textfield.container.attr("value"));
```
Comment (jobara) [inline tests/component-tests/textfieldControl/js/TextfieldStepperTests.js:92]: @greatislander related to @amb26 comment above about the value of the input field, this should like get the actual value using something like `that.textfield.container.val()`.

File: src/components/textfieldControl/css/TextfieldStepper.css (hunk 1 above)
Comment (greatislander) [reply src/components/textfieldControl/css/TextfieldStepper.css:26]: As the `-moz` prefix would suggest, it's needed for Firefox to display this input without the default stepper controls. The route to eliminating it would be for Firefox to provide an alternative method of doing this that doesn't require a browser prefix.

File: src/framework/preferences/css/sass/PrefsEditor.scss (hunk 3 above)
Comment (greatislander) [reply src/framework/preferences/css/sass/PrefsEditor.scss:208]: The prefs framework and UIO demos don't include the TextfieldStepper stylesheet— it appears that for some historical reason, those styles were duplicated in the Prefs Framework stylesheet.

File: src/components/textfieldControl/js/TextfieldStepper.js (hunk 2 above)
Comment (greatislander) [reply src/components/textfieldControl/js/TextfieldStepper.js]: It appears that this listener can indeed be removed as it's handled in the textfield component. Resolved in https://github.com/fluid-project/infusion/pull/1049/commits/fe7b771984c45608edce74a6485453ea5cff0b5d.

File: tests/component-tests/textfieldControl/js/TextfieldStepperTests.js (hunk 4 above)
Comment (greatislander) [reply tests/component-tests/textfieldControl/js/TextfieldStepperTests.js]: I've added a test to do this in 06f3533.

File: src/framework/preferences/css/sass/PrefsEditor.scss (hunk 3 above)
Comment (jobara) [reply src/framework/preferences/css/sass/PrefsEditor.scss:208]: Hmm... maybe it was to simplify the imports, not sure.

File: src/components/textfieldControl/css/TextfieldStepper.css [hunk 5]
```diff
@@ -23,13 +23,20 @@
...
     font-family: inherit;
     font-size: 100%;
     margin: 0 0.5rem;
     text-align: center;
     width: 7rem;
 }
 
+.fl-textfieldStepper .fl-textfieldStepper-textField::-webkit-inner-spin-button,
+.fl-textfieldStepper .fl-textfieldStepper-textField::-webkit-outer-spin-button {
+    appearance: none;
```
Comment (jobara) [inline src/components/textfieldControl/css/TextfieldStepper.css:36]: It appears this needs to be `-webkit-appearance: none;` for it to work on Safari.

How it looks as it currently is:
Number input in Safari

Comment (greatislander) [reply src/components/textfieldControl/css/TextfieldStepper.css]: Resolved in 33aa7dc.

**Comment by jobara** [comment]
The input fields are generally smaller now. Ideally the input field should be the same height as the stepper buttons on the left and right. (It's a bit hard to parse, but you can see some of the older [designs](https://wiki.fluidproject.org/display/fluid/%28Floe%29+User+Interface+%28Learner%29+Options+Mobile+and+Responsive+Design) as a reference and [FLUID-5536](https://issues.fluidproject.org/browse/FLUID-5536))

e.g. from the [build site](https://build-infusion.fluidproject.org/demos/uioptions/) in Safari
Screen Shot 2021-05-13 at 9 44 02 AM

from Safari with the changes from this PR
Screen Shot 2021-05-13 at 9 45 14 AM

Other browsers are displaying the input size smaller as well, and may be partially related to removing normalize in [FLUID-6607](https://issues.fluidproject.org/browse/FLUID-6607)

**Comment by jobara** [comment]
When using VoiceOver on macOS with Chrome and MS Edge, keyboard support doesn't work quite right. If you navigate to the input field, pressing the <kbd>↑</kbd> or <kbd>↓</kbd> shifts focus to the "+" button on the initial press. This prevents the arrow keys from incrementing/decrementing the values and the button is also not interactive with the keyboard. If you <kbd>shift</kbd>+<kbd>tab</kbd> back, you can use the <kbd>↑</kbd> or <kbd>↓</kbd> keys as expected.

**Comment by greatislander** [comment]
> The input fields are generally smaller now.

Resolved in 1eab77d.

File: src/components/textfieldControl/js/TextfieldStepper.js [hunk 6]
```diff
@@ -74,18 +79,6 @@ https://github.com/fluid-project/infusion/raw/main/Infusion-LICENSE.txt
                             listener: "fluid.textfieldStepper.bindKeyEvent",
                             // down arrow === 40
                             args: ["{that}.container", "keydown", 40, "{textfieldStepper}.decrease"]
-                        },
```
Comment (jobara) [inline src/components/textfieldControl/js/TextfieldStepper.js:77]: This probably aren't necessary anymore and may be causing the issue with the arrow keys that I mentioned in VO and Chrome/MS Edge.

Comment (greatislander) [reply src/components/textfieldControl/js/TextfieldStepper.js:77]: Resolved in 6d0ef96.

File: tests/component-tests/textfieldControl/js/TextfieldStepperTests.js [hunk 7]
```diff
@@ -148,17 +148,17 @@ https://github.com/fluid-project/infusion/raw/main/Infusion-LICENSE.txt
                         changeEvent: "{stepper}.applier.modelChanged"
                     }, {
                         // decrease to minimum
-                        jQueryTrigger: fluid.tests.textfieldControl.createKeyEvent("keydown", 40),
-                        element: "{stepper}.dom.textfield"
+                        jQueryTrigger: "click",
+                        element: "{stepper}.dom.decreaseButton"
```
Comment (jobara) [inline tests/component-tests/textfieldControl/js/TextfieldStepperTests.js:152]: Did the keyboard interaction tests stop working after the change? It would be good to test if the up/down arrow keys still work to increment and decrement the model values.

Comment (greatislander) [reply tests/component-tests/textfieldControl/js/TextfieldStepperTests.js]: I reverted this change and now the tests hang locally (but on the TextFieldSlider tests, not this one). I'm going to see if they run in GitHub actions.

Comment (greatislander) [reply tests/component-tests/textfieldControl/js/TextfieldStepperTests.js]: Looks like they don't. No output, no messages.

Comment (jobara) [reply tests/component-tests/textfieldControl/js/TextfieldStepperTests.js]: The "No output" is probably because the IoC testing framework is waiting for an event that never fires. In the browser you get an error in the console for this, but jqUnit is kind of stuck. I filed a [FLUID-6622](https://issues.fluidproject.org/browse/FLUID-6622) about this. 

Anyways, the error is probably because the browser is ignore the synthetic keyboard interaction. There are some interactions that the browser requires to be user directed to actually work. It seems that this might be one of those cases. In that case, we won't be able to test the keyboard interaction, and you can go back to using the mouse click on the buttons.

Thanks for looking into this though.

Comment (greatislander) [reply tests/component-tests/textfieldControl/js/TextfieldStepperTests.js]: That was my instinct based on a previous conversation we'd had which was why I removed it. I'll revert the revert.
